### PR TITLE
[gdal] Fix dependency hdf5

### LINKED ISSUE #25645: [gdal] build failure
**Host Environment**
- OS: windows
- Compiler: MSVC/14.32.31326 (i.e. VS 2022)

**To Reproduce**
`./vcpkg.exe install gdal --triplet x64-windows`

**Failure logs**
`-- Configuring x64-windows
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:96 (message):
    Command failed: "C:/Program Files/Microsoft Visual Studio/2022/Professional/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -v
    Working Directory: C:/build/vcpkg/buildtrees/gdal/x64-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      C:\build\vcpkg\buildtrees\gdal\config-x64-windows-CMakeCache.txt.log
      C:\build\vcpkg\buildtrees\gdal\config-x64-windows-CMakeCache.txt.log
      C:\build\vcpkg\buildtrees\gdal\config-x64-windows-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:238 (vcpkg_execute_required_process)
  ports/gdal/portfile.cmake:62 (vcpkg_cmake_configure)
  scripts/ports.cmake:146 (include)

error: building gdal:x64-windows failed with: BUILD_FAILED`

**Additional context**
Builds with default-features and hda5 removed, i.e. gdal[core,curl,geos,libspatialite,netcdf,postgresql,recommended-features]
CMake is 3.24.0-rc3

[config-x64-windows-out.log](https://github.com/microsoft/vcpkg/files/9071293/config-x64-windows-out.log)
[config-x64-windows-CMakeCache.txt.log](https://github.com/microsoft/vcpkg/files/9071294/config-x64-windows-CMakeCache.txt.log)

### PR DESCRIPTION
**Pull request by JackBoosY**
Related: https://github.com/microsoft/vcpkg/pull/22392#issuecomment-1178629755

### CODE REVIEW

**Comment by JackBoosY** [comment]
cc @dg0yt

File: ports/gdal/fix-dependency-hdf5.patch [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+diff --git a/frmts/hdf5/CMakeLists.txt b/frmts/hdf5/CMakeLists.txt
+index 4b85797..919f061 100644
+--- a/frmts/hdf5/CMakeLists.txt
++++ b/frmts/hdf5/CMakeLists.txt
+@@ -24,7 +24,7 @@ if (WIN32)
+   target_compile_definitions(gdal_HDF5 PRIVATE -DWIN32)
+ endif ()
+ target_include_directories(gdal_HDF5 SYSTEM PRIVATE ${HDF5_INCLUDE_DIRS})
+-gdal_target_link_libraries(gdal_HDF5 PRIVATE ${HDF5_C_LIBRARIES})
++target_link_libraries(gdal_HDF5 PRIVATE ${HDF5_C_LIBRARIES})
```
Comment (JackBoosY) [inline ports/gdal/fix-dependency-hdf5.patch:26]:
```cmake
function(gdal_target_interfaces _TARGET)
    foreach (_LIB IN ITEMS ${ARGN})
        if (TARGET ${_LIB})
            get_property(_res TARGET ${_LIB} PROPERTY INTERFACE_INCLUDE_DIRECTORIES)
            if (_res)
                target_include_directories(${_TARGET} PRIVATE ${_res})
            endif ()
            get_property(_res TARGET ${_LIB} PROPERTY INTERFACE_COMPILE_DEFINITIONS)
            if (_res)
                target_compile_definitions(${_TARGET} PRIVATE ${_res})
            endif ()
            get_property(_res TARGET ${_LIB} PROPERTY INTERFACE_COMPILE_OPTIONS)
            if (_res)
                target_compile_options(${_TARGET} PRIVATE ${_res})
            endif ()
        endif ()
    endforeach ()
endfunction()
```

```cmake
function (gdal_add_private_link_libraries)
  get_property(tmp GLOBAL PROPERTY gdal_private_link_libraries)
  foreach (arg ${ARGV})
    set(tmp ${tmp} ${arg})
  endforeach ()
  set_property(GLOBAL PROPERTY gdal_private_link_libraries ${tmp})
endfunction (gdal_add_private_link_libraries)
```

So cmake will not add the `INTERFACE_INCLUDE_DIRECTORIES`.

Comment (dg0yt) [reply ports/gdal/fix-dependency-hdf5.patch:26]: > So cmake will not add the `INTERFACE_INCLUDE_DIRECTORIES`.

There is `target_include_directories(gdal_HDF5 SYSTEM PRIVATE ${HDF5_INCLUDE_DIRS})` (line 24), and this is enough.
GDAL is meant to use CMake's `FindHDF5.cmake`. And this passed CI in GDAL and in vcpkg. Multiple times.

This doesn't mean there could be broken configuration. But you have to clearly identify this configuration (for a clean start). And you proabbly have to find and fix it in the vicinity of port hdf5.

And there configuration options which might remove the need for patching. It is just not clear enough what you try to fix.

Comment (dg0yt) [reply ports/gdal/fix-dependency-hdf5.patch:26]: No x64-windows here, but there: #25645

Comment (JackBoosY) [reply ports/gdal/fix-dependency-hdf5.patch:26]: The problem is that neither vcpkg-cmake-wrapper nor hdf5-config.cmake provide `HDF5_INCLUDE_DIRS`.

**Comment by JackBoosY** [comment]
In favor of https://github.com/microsoft/vcpkg/pull/25646

**Pull request closed without merging**